### PR TITLE
:sparkles: Introduces flipCoin()

### PR DESCRIPTION
**Pull request by canassa**
`GenerateRnd(2) != 0` and `GenerateRnd(2) == 0` were being used everwhere so I created a new function specifically for that.

The `flipCoin` name is debatable but I think that it shows a better intent than `randomBoolean`

### CODE REVIEW

**Comment by canassa** [comment]
Hmm, the tests broke because the existing code is using `GenerateRnd(2) != 0` and `GenerateRnd(2) == 0` interchangeably

File: Source/engine/random.cpp [hunk 1]
```diff
@@ -51,4 +51,9 @@ int32_t GenerateRnd(int32_t v)
 	return AdvanceRndSeed() % v;
 }
 
+bool flipCoin()
+{
+	return GenerateRnd(2) == 0;
```
Comment (AJenbo) [inline Source/engine/random.cpp:56]: There are 7 cases of this and 40 cases of the other variant, so change this to the common one and change the 7 other cases to use `!flipCoin()`

```suggestion
	return GenerateRnd(2) != 0;
```

**Comment by AJenbo** [comment]
> I checked the tests and I think I could fix them, but let me know first if this change is welcome or not before I spend time on it. I am not sure if there are impacts on changing the level generation like that. With this change, a different dungeon will be generated for the same seed

The tests for levels and item generation is there to make sure that we stay compatible with the original game. It is critical that they do not change.

**Comment by AJenbo** [comment]
> The `flipCoin` name is debatable but I think that it shows a better intent than `randomBoolean`

Fun fact I have Cuprolaminophobia :(

File: Source/engine/random.cpp (hunk 1 above)
Comment (AJenbo) [reply Source/engine/random.cpp]: Alright just that one line fixed the 5 failing item generation tests :)

File: Source/objects.cpp [hunk 2]
```diff
@@ -987,7 +987,7 @@ void DeleteObject(int oi, int i)
 
 void AddChest(int i, int t)
 {
-	if (GenerateRnd(2) == 0)
+	if (flipCoin())
```
Comment (AJenbo) [inline Source/objects.cpp:990]: I'm guessing this wasn't deliberate as it is the only one you didn't flip :)
```suggestion
	if (!flipCoin())
```

File: Source/engine/random.cpp [hunk 3]
```diff
@@ -51,4 +51,9 @@ int32_t GenerateRnd(int32_t v)
 	return AdvanceRndSeed() % v;
 }
 
+bool flipCoin()
```
Comment (AJenbo) [inline Source/engine/random.cpp:54]: I wasn't able to come up with a better name, but being a global the casing is wrong:
```suggestion
bool FlipCoin()
```

File: Source/levels/drlg_l1.cpp [hunk 4]
```diff
@@ -954,10 +954,10 @@ void GenerateRoom(int x, int y, int w, int h, int dir)
...
 void FirstRoom()
 {
-	VerticalLayout = GenerateRnd(2) == 0;
-	HasChamber1 = (GenerateRnd(2) != 0);
-	HasChamber2 = (GenerateRnd(2) != 0);
-	HasChamber3 = (GenerateRnd(2) != 0);
+	VerticalLayout = !FlipCoin();
+	HasChamber1 = FlipCoin();
+	HasChamber2 = FlipCoin();
+	HasChamber3 = FlipCoin();
```
Comment (AJenbo) [inline Source/levels/drlg_l1.cpp:960]: All though it seams strange that VerticalLayout uses an inverted check, I think you will agree that it's better then the original code :)

https://github.com/diasurgical/devilution/blob/master/Source/drlg_l1.cpp#L1486

File: Source/engine/random.cpp [hunk 5]
```diff
@@ -51,4 +51,9 @@ int32_t GenerateRnd(int32_t v)
 	return AdvanceRndSeed() % v;
 }
 
+bool FlipCoin()
+{
+	return GenerateRnd(2) != 0;
+}
```
Comment (ephphatha) [inline Source/engine/random.cpp:57]: Structuring this check to be `== 0` is more natural, especially when dealing with frequencies other than 1 in 2. To fully address #2193 you could make this function take a frequency parameter with a default of 2. Then you can replace calls to GenerateRnd() like items.cpp:1084,1156,1157,1167,1224,1600, objects.cpp:504,538,542,546,550, etc.

File: Source/levels/drlg_l1.cpp (hunk 4 above)
Comment (AJenbo) [reply Source/levels/drlg_l1.cpp:960]: There we go: https://github.com/diasurgical/devilutionX/pull/4912

File: Source/engine/random.cpp (hunk 5 above)
Comment (AJenbo) [reply Source/engine/random.cpp:57]: See https://github.com/diasurgical/devilutionX/pull/4908#discussion_r916160846 for why we flipped it. Not that we couldn't just add a lot of `!` instead.

Comment (ephphatha) [reply Source/engine/random.cpp:57]: Opened #4913 with the result of inverting the condition, can adjust most of the callers to suit :D

**Pull request merged**